monai/losses/image_dissimilarity.py

```diff
@@ -0,0 +1,155 @@
+from monai.utils import LossReduction
+
+
+class LocalNormalizedCrossCorrelationLoss(_Loss):
+    """
+    Local squared zero-normalized cross-correlation.
+    The loss is based on a moving kernel/window over the y_true/y_pred,
+    within the window the square of zncc is calculated.
+    The kernel can be a rectangular / triangular / gaussian window.
+    The final loss is the averaged loss over all windows.
+
+    Adapted from:
+        https://github.com/voxelmorph/voxelmorph/blob/legacy/src/losses.py
+        DeepReg (https://github.com/DeepRegNet/DeepReg)
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        ndim: int = 3,
+        kernel_size: int = 9,
+        kernel_type: str = "rectangular",
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+        smooth_nr: float = 1e-7,
+        smooth_dr: float = 1e-7,
+    ) -> None:
+        """
+        Args:
+            in_channels: number of input channels
+            ndim: number of spatial ndimensions, {``1``, ``2``, ``3``}. Defaults to 3.
+            kernel_size: kernel spatial size, must be odd.
+            kernel_type: {``"rectangular"``, ``"triangular"``, ``"gaussian"``}. Defaults to ``"rectangular"``.
+            reduction: {``"none"``, ``"mean"``, ``"sum"``}
+                Specifies the reduction to apply to the output. Defaults to ``"mean"``.
+
+                - ``"none"``: no reduction will be applied.
+                - ``"mean"``: the sum of the output will be divided by the number of elements in the output.
+                - ``"sum"``: the output will be summed.
+            smooth_nr: a small constant added to the numerator to avoid zero.
+            smooth_dr: a small constant added to the denominator to avoid nan.
+        """
+        super(LocalNormalizedCrossCorrelationLoss, self).__init__(reduction=LossReduction(reduction).value)
+        self.in_channels = in_channels
+
+        self.ndim = ndim
+        if self.ndim not in [1, 2, 3]:
+            raise ValueError(f"Unsupported ndim: {self.ndim}-d, only 1-d, 2-d, and 3-d inputs are supported")
+
+        self.kernel_size = kernel_size
+        if self.kernel_size % 2 == 0:
+            raise ValueError(f"kernel_size must be odd, got {self.kernel_size}")
+
+        if kernel_type == "rectangular":
+            self.kernel = self.make_rectangular_kernel()
+        elif kernel_type == "triangular":
+            self.kernel = self.make_triangular_kernel()
+        elif kernel_type == "gaussian":
+            self.kernel = self.make_gaussian_kernel()
+        else:
+            raise ValueError(
+                f'Unsupported kernel_type: {kernel_type}, available options are ["rectangular", "triangular", "gaussian"].'
+            )
+
+        self.kernel_vol = torch.sum(self.kernel) ** self.ndim
+        self.smooth_nr = float(smooth_nr)
+        self.smooth_dr = float(smooth_dr)
+
+    def make_rectangular_kernel(self):
```

Comment:
These kernel functions requires only the argument `kernel_size`.

I suggest to move these functions outside of the class, or mark them as `@staticmethod`.

Reply:
Done